// lifemaxxing-shared/rankLadder.js
'use strict';

// Rank ladder + Chad/Chad Light consistency gating. XP alone is necessary
// but not sufficient for the top two ranks - see RANKS[*].consistency.

const RANKS = [
  { key: 'LOW_TIER_NORMIE_1', label: 'Low Tier Normie 1', xpRequired: 0 },
  { key: 'LOW_TIER_NORMIE_2', label: 'Low Tier Normie 2', xpRequired: 100 },
  { key: 'LOW_TIER_NORMIE_3', label: 'Low Tier Normie 3', xpRequired: 250 },
  { key: 'MID_TIER_NORMIE_1', label: 'Mid Tier Normie 1', xpRequired: 500 },
  { key: 'MID_TIER_NORMIE_2', label: 'Mid Tier Normie 2', xpRequired: 900 },
  { key: 'MID_TIER_NORMIE_3', label: 'Mid Tier Normie 3', xpRequired: 1400 },
  { key: 'HIGH_TIER_NORMIE_1', label: 'High Tier Normie 1', xpRequired: 2100 },
  { key: 'HIGH_TIER_NORMIE_2', label: 'High Tier Normie 2', xpRequired: 3000 },
  { key: 'HIGH_TIER_NORMIE_3', label: 'High Tier Normie 3', xpRequired: 4200 },
  {
    key: 'CHAD_LIGHT',
    label: 'Chad Light',
    xpRequired: 8000,
    consistency: {
      activeDaysWindow: 30,
      activeDaysRequired: 21,
      qualifyingStreaksRequired: 2,
      qualifyingStreakMinDays: 14,
    },
  },
  {
    key: 'CHAD',
    label: 'Chad',
    xpRequired: 18000,
    consistency: {
      activeDaysWindow: 60,
      activeDaysRequired: 45,
      qualifyingStreaksRequired: 3,
      qualifyingStreakMinDays: 30,
      eliteWeeksRequired: 4,
    },
  },
];

const ELITE_WEEK_MIN_XP = 600;
const ELITE_WEEK_MIN_CATEGORIES = 3;

function meetsConsistency(rank, stats) {
  const req = rank.consistency;
  if (!req) return true;

  const activeDaysOk = req.activeDaysWindow === 30
    ? stats.activeDaysLast30 >= req.activeDaysRequired
    : stats.activeDaysLast60 >= req.activeDaysRequired;

  const qualifyingStreaksOk = (stats.categoryStreakDays || [])
    .filter((days) => days >= req.qualifyingStreakMinDays).length >= req.qualifyingStreaksRequired;

  const eliteWeeksOk = !req.eliteWeeksRequired || (stats.eliteWeeksCount || 0) >= req.eliteWeeksRequired;

  return activeDaysOk && qualifyingStreaksOk && eliteWeeksOk;
}

/**
 * Walks the ladder top-down and returns the highest rank the user actually
 * qualifies for: XP threshold met AND (if applicable) consistency
 * requirements met. A user who hits 8,000 XP but lacks the active-days
 * requirement stays at High Tier Normie 3, not Chad Light.
 *
 * @param {object} stats
 * @param {number} stats.totalXP
 * @param {number} stats.activeDaysLast30
 * @param {number} stats.activeDaysLast60
 * @param {number[]} stats.categoryStreakDays - current streak length per category
 * @param {number} stats.eliteWeeksCount
 */
function computeRank(stats) {
  for (let i = RANKS.length - 1; i >= 0; i -= 1) {
    const rank = RANKS[i];
    if (stats.totalXP >= rank.xpRequired && meetsConsistency(rank, stats)) {
      return { ...rank, rankIndex: i };
    }
  }
  return { ...RANKS[0], rankIndex: 0 };
}

function isEliteWeek({ weeklyXP, categoriesUsedThisWeek }) {
  return weeklyXP >= ELITE_WEEK_MIN_XP && categoriesUsedThisWeek >= ELITE_WEEK_MIN_CATEGORIES;
}

module.exports = {
  RANKS,
  ELITE_WEEK_MIN_XP,
  ELITE_WEEK_MIN_CATEGORIES,
  computeRank,
  isEliteWeek,
};
