// lifemaxxing-shared/xpRules.js
'use strict';

// Single source of truth for XP math. Any change to the game's balance
// (base XP, multiplier curve, caps) happens here and nowhere else.

const CATEGORIES = {
  FITNESS: 'FITNESS',
  SCREEN_DISCIPLINE: 'SCREEN_DISCIPLINE',
  FOCUS: 'FOCUS',
  PERSONAL_GOALS: 'PERSONAL_GOALS',
  REFLECTION: 'REFLECTION',
  SPIRITUAL: 'SPIRITUAL',
};

// Optional categories the user can toggle on/off in Category Setup.
const OPTIONAL_CATEGORIES = new Set([CATEGORIES.REFLECTION, CATEGORIES.SPIRITUAL]);

const BASE_XP = {
  [CATEGORIES.FITNESS]: 30,
  [CATEGORIES.SCREEN_DISCIPLINE]: 25,
  [CATEGORIES.FOCUS]: 10, // per 25-minute block
  [CATEGORIES.PERSONAL_GOALS]: 10,
  [CATEGORIES.REFLECTION]: 10,
  [CATEGORIES.SPIRITUAL]: 15, // doc allows 10-20; 15 is the MVP constant
};

const CATEGORY_DAILY_CAP = 55;
const TOTAL_DAILY_CAP = 125;

// Multiplier curve, confirmed source of truth: caps at 1.30x on day 7.
// The streak *count* can keep climbing past 7 for badges/status, but the
// XP multiplier never exceeds 1.30x.
function getMultiplierForStreak(streakDays) {
  if (streakDays >= 7) return 1.30;
  if (streakDays >= 5) return 1.20;
  if (streakDays >= 3) return 1.10;
  if (streakDays >= 2) return 1.05;
  return 1.00; // 0-1 day streak
}

function isValidCategory(categoryId) {
  return Object.prototype.hasOwnProperty.call(BASE_XP, categoryId);
}

/**
 * Computes the XP to award for one completed action, applying the
 * calculation order from the reference doc:
 *   base XP -> streak multiplier -> per-category daily cap -> total daily cap -> round
 *
 * @param {object} params
 * @param {string} params.categoryId
 * @param {number} params.streakDaysAfterThisCompletion - the category streak count INCLUDING today's completion
 * @param {number} params.categoryXPAlreadyToday - XP already awarded in this category today, before this completion
 * @param {number} params.totalXPAlreadyToday - XP already awarded across all categories today, before this completion
 * @returns {{ baseXP: number, multiplier: number, xpAfterMultiplier: number, finalXPAwarded: number, categoryCapApplied: boolean, dailyCapApplied: boolean }}
 */
function calculateAwardedXP({
  categoryId,
  streakDaysAfterThisCompletion,
  categoryXPAlreadyToday,
  totalXPAlreadyToday,
}) {
  if (!isValidCategory(categoryId)) {
    throw new Error(`Unknown categoryId: ${categoryId}`);
  }

  const baseXP = BASE_XP[categoryId];
  const multiplier = getMultiplierForStreak(streakDaysAfterThisCompletion);
  const xpAfterMultiplier = baseXP * multiplier;

  const categoryRemaining = Math.max(0, CATEGORY_DAILY_CAP - categoryXPAlreadyToday);
  const afterCategoryCap = Math.min(xpAfterMultiplier, categoryRemaining);
  const categoryCapApplied = afterCategoryCap < xpAfterMultiplier;

  const totalRemaining = Math.max(0, TOTAL_DAILY_CAP - totalXPAlreadyToday);
  const afterDailyCap = Math.min(afterCategoryCap, totalRemaining);
  const dailyCapApplied = afterDailyCap < afterCategoryCap;

  const finalXPAwarded = Math.round(afterDailyCap);

  return {
    baseXP,
    multiplier,
    xpAfterMultiplier,
    finalXPAwarded,
    categoryCapApplied,
    dailyCapApplied,
  };
}

module.exports = {
  CATEGORIES,
  OPTIONAL_CATEGORIES,
  BASE_XP,
  CATEGORY_DAILY_CAP,
  TOTAL_DAILY_CAP,
  getMultiplierForStreak,
  isValidCategory,
  calculateAwardedXP,
};
